File: qlang-interpreter/src/qlang/runtime/values.ts
import { Callable } from "./callable"

export type ValueType =
    | 'null'
    | 'number'
    | 'boolean'
    | 'string'
    | 'break'
    | 'continue'
    | 'return'
    | 'array'
    | 'function'

export interface RuntimeValue {
    type: ValueType
}

export interface BreakValue extends RuntimeValue {
    type: 'break'
}

export function MK_BREAK(): BreakValue {
    return { type: 'break' }
}

export interface ContinueValue extends RuntimeValue {
    type: 'continue'
}

export function MK_CONTINUE(): ContinueValue {
    return { type: 'continue' }
}

type AlgebraicType = 'number' | 'string' | 'boolean' | 'null' | 'array' | 'function' | 'return'
type AlgebraicValueType = number | string | boolean | null | AlgebraicValue[] | Callable

export interface AlgebraicValue extends RuntimeValue {
    type: AlgebraicType
    value: number | string | boolean | null | AlgebraicValue[] | Callable
}

export function MK_ALGEBRAIC(value: AlgebraicValueType): AlgebraicValue {
    if (typeof value === 'number') {
        return MK_NUMBER(value)
    }
    if (typeof value === 'string') {
        return MK_STRING(value)
    }
    if (typeof value === 'boolean') {
        return MK_BOOLEAN(value)
    }
    if (value === null) {
        return MK_NULL()
    }
    if (Array.isArray(value)) {
        return MK_ARRAY(value)
    }
    if ('Arity' in value) {
        return MK_FUNCTION(value)
    }
    throw new Error(`Unsupported algebraic value: ${value}`)
}

export interface FunctionValue extends AlgebraicValue {
    type: 'function'
    value: Callable
}

export function MK_FUNCTION(value: Callable): FunctionValue {
    return { type: 'function', value }
}

export interface ReturnValue extends AlgebraicValue {
    type: 'return'
}

export function MK_RETURN(returnValue: AlgebraicValue): ReturnValue {
    return { type: 'return', value: returnValue.value }
}

export interface ArrayValue extends AlgebraicValue {
    type: 'array'
    value: AlgebraicValue[]
}

export function MK_ARRAY(value: AlgebraicValue[]): ArrayValue {
    return { type: 'array', value }
}

export interface NullValue extends AlgebraicValue {
    type: 'null'
    value: null
}

export function MK_NULL(): NullValue {
    return { type: 'null', value: null }
}

export interface NumberValue extends AlgebraicValue {
    type: 'number'
    value: number
}

export function MK_NUMBER(value: number): NumberValue {
    return { type: 'number', value }
}

export interface BooleanValue extends AlgebraicValue {
    type: 'boolean'
    value: boolean
}

export function MK_BOOLEAN(value: boolean): BooleanValue {
    return { type: 'boolean', value }
}

export interface StringValue extends AlgebraicValue {
    type: 'string'
    value: string
}

export function MK_STRING(value: string): StringValue {
    return { type: 'string', value }
}

File: interpreter/src/runtime/values.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import TYPE_CHECKING


if TYPE_CHECKING:
    from src.runtime.callable import Callable


class AlgebraicType(Enum):
    NULL = "null"
    NUMBER = "number"
    BOOLEAN = "boolean"
    STRING = "string"
    BREAK = "break"
    CONTINUE = "continue"
    RETURN = "return"
    ARRAY = "array"
    FUNCTION = "function"


@dataclass
class RuntimeValue:
    type: AlgebraicType


@dataclass
class BreakValue(RuntimeValue):
    type: AlgebraicType = AlgebraicType.BREAK


@dataclass
class ContinueValue(RuntimeValue):
    type: AlgebraicType = AlgebraicType.CONTINUE


@dataclass
class AlgebraicValue(RuntimeValue):
    value: None | float | bool | str | list[AlgebraicValue] | Callable


@dataclass
class NullValue(AlgebraicValue):
    type: AlgebraicType = AlgebraicType.NULL
    value: None = None


@dataclass
class NumberValue(AlgebraicValue):
    type: AlgebraicType = AlgebraicType.NUMBER
    value: float = 0.0


@dataclass
class BooleanValue(AlgebraicValue):
    type: AlgebraicType = AlgebraicType.BOOLEAN
    value: bool = False


@dataclass
class StringValue(AlgebraicValue):
    type: AlgebraicType = AlgebraicType.STRING
    value: str = ""


@dataclass
class ReturnValue(AlgebraicValue):  # type: ignore[misc]
    type: AlgebraicType = AlgebraicType.RETURN


@dataclass
class ArrayValue(AlgebraicValue):
    type: AlgebraicType = AlgebraicType.ARRAY
    value: list[AlgebraicValue] = []


@dataclass
class FunctionValue(AlgebraicValue):
    value: Callable  # type: ignore[misc]
    type: AlgebraicType = AlgebraicType.FUNCTION


def MK_BREAK() -> BreakValue:
    return BreakValue(type=AlgebraicType.BREAK)


def MK_CONTINUE() -> ContinueValue:
    return ContinueValue(type=AlgebraicType.CONTINUE)


def MK_NULL() -> NullValue:
    return NullValue(type=AlgebraicType.NULL, value=None)


def MK_NUMBER(value: float) -> NumberValue:
    return NumberValue(type=AlgebraicType.NUMBER, value=value)


def MK_BOOLEAN(value: bool) -> BooleanValue:
    return BooleanValue(type=AlgebraicType.BOOLEAN, value=value)


def MK_STRING(value: str) -> StringValue:
    return StringValue(type=AlgebraicType.STRING, value=value)


def MK_RETURN(return_value: AlgebraicValue) -> ReturnValue:
    return ReturnValue(type=AlgebraicType.RETURN, value=return_value.value)


def MK_ARRAY(value: list[AlgebraicValue]) -> ArrayValue:
    return ArrayValue(type=AlgebraicType.ARRAY, value=value)


def MK_FUNCTION(value: Callable) -> FunctionValue:
    return FunctionValue(type=AlgebraicType.FUNCTION, value=value)


def MK_ALGEBRAIC_VALUE(
    value: None | float | bool | str | list[AlgebraicValue] | Callable,
) -> AlgebraicValue:
    if value is None:
        return MK_NULL()
    if value is float:
        f_value: float = value  # type: ignore
        return MK_NUMBER(f_value)
    if value is bool:
        return MK_BOOLEAN(bool(value))
    if value is str:
        return MK_STRING(str(value))
    if value is list:
        arr: list[AlgebraicValue] = value  # type: ignore
        return MK_ARRAY(arr.copy())
    if value is Callable:
        func: Callable = value  # type: ignore
        return MK_FUNCTION(func)
    return MK_NULL()
